### ISSUE TITLE
[Bug] _fwd_kernel_ep_scatter_1 may lead to illegal memory access.

### ISSUE THREAD

**Issue by HPC4AI** [sgl-project/sglang#31929]
### Checklist

- [ ] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Problem

I found that the implementation of the `_fwd_kernel_ep_scatter_1` function in the file `sglang/python/sglang/kernels/ops/moe/ep_moe_kernels.py` may lead to illegal memory access. The source code is as follows:

```
@triton.jit
def _fwd_kernel_ep_scatter_1(
    num_recv_tokens_per_expert,
    expert_start_loc,
    m_indices,
    num_experts: tl.constexpr,
    BLOCK_E: tl.constexpr,
    BLOCK_EXPERT_NUM: tl.constexpr,
):
    cur_expert = tl.program_id(0)

    offset_cumsum = tl.arange(0, BLOCK_EXPERT_NUM)
    tokens_per_expert = tl.load(
        num_recv_tokens_per_expert + offset_cumsum,
        mask=offset_cumsum < num_experts,
        other=0,
    )
    cumsum = tl.cumsum(tokens_per_expert) - tokens_per_expert
    tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)
    
    cur_expert_start = tl.load(expert_start_loc + cur_expert)
    cur_expert_token_num = tl.load(num_recv_tokens_per_expert + cur_expert)

    m_indices_start_ptr = m_indices + cur_expert_start
    off_expert = tl.arange(0, BLOCK_E)

    for start_m in tl.range(0, cur_expert_token_num, BLOCK_E, num_stages=4):
        tl.store(
            m_indices_start_ptr + start_m + off_expert,
            cur_expert,
        )
```

Specifically,

`tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)`

only a subset of threads within the CTA perform the write, 

`cur_expert_start = tl.load(expert_start_loc + cur_expert)`

while all threads in the CTA perform the read. There is no synchronization between these operations, which may cause incorrect values of `cur_expert_start`, leading to illegal memory access when accessing `m_indices_start_ptr`.

During end-to-end inference execution, this issue may not immediately trigger an error. However, the CUDA illegal memory access can corrupt the CUDA context, preventing the DeepEP normal kernel from being issued to the CUDA stream. This further results in the CPU-side exception `"DeepEP error: CPU recv timeout"`, and only then does the inference engine log report an "illegal memory access" error. The issue cannot be reproduced when `CUDA_LAUNCH_BLOCKING` is enabled; I was only able to accurately pinpoint the root cause by incrementally removing unrelated code.

### How to fix?

The first modification approach is to avoid reading data from `expert_start_loc` and instead compute `cur_expert_start` directly. Ref to https://github.com/sgl-project/sglang/pull/31930

```
@triton.jit
def ep_scatter_prefix_sum_kernel(
    num_recv_tokens_per_expert,
    expert_start_loc,
    m_indices,
    num_experts: tl.constexpr,
    BLOCK_E: tl.constexpr,
    BLOCK_EXPERT_NUM: tl.constexpr,
):
    """Get the current expert offset by reducing the values already in registers."""
    cur_expert = tl.program_id(0)

    offset_cumsum = tl.arange(0, BLOCK_EXPERT_NUM)
    tokens_per_expert = tl.load(
        num_recv_tokens_per_expert + offset_cumsum,
        mask=offset_cumsum < num_experts,
        other=0,
    )
    cumsum = tl.cumsum(tokens_per_expert) - tokens_per_expert
    tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)

    cur_expert_start = tl.sum(
        tl.where(offset_cumsum < cur_expert, tokens_per_expert, 0),
        axis=0,
    )
    cur_expert_token_num = tl.load(num_recv_tokens_per_expert + cur_expert)

    m_indices_start_ptr = m_indices + cur_expert_start
    off_expert = tl.arange(0, BLOCK_E)

    for start_m in tl.range(0, cur_expert_token_num, BLOCK_E, num_stages=4):
        tl.store(m_indices_start_ptr + start_m + off_expert, cur_expert)
```

The second approach is to use `tl.debug_barrier()`.

```
def _fwd_kernel_ep_scatter_1(
    num_recv_tokens_per_expert,
    expert_start_loc,
    m_indices,
    num_experts: tl.constexpr,
    BLOCK_E: tl.constexpr,
    BLOCK_EXPERT_NUM: tl.constexpr,
):
    cur_expert = tl.program_id(0)

    offset_cumsum = tl.arange(0, BLOCK_EXPERT_NUM)
    tokens_per_expert = tl.load(
        num_recv_tokens_per_expert + offset_cumsum,
        mask=offset_cumsum < num_experts,
        other=0,
    )
    cumsum = tl.cumsum(tokens_per_expert) - tokens_per_expert
    tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)
    tl.debug_barrier()
    
    cur_expert_start = tl.load(expert_start_loc + cur_expert)
    cur_expert_token_num = tl.load(num_recv_tokens_per_expert + cur_expert)

    m_indices_start_ptr = m_indices + cur_expert_start
    off_expert = tl.arange(0, BLOCK_E)

    for start_m in tl.range(0, cur_expert_token_num, BLOCK_E, num_stages=4):
        tl.store(
            m_indices_start_ptr + start_m + off_expert,
            cur_expert,
        )
```

### Reproduction

 Dump PTX Code

```
#!/usr/bin/env python3
"""Compile and dump PTX for two variants of the EP scatter prefix kernel."""

import argparse
from pathlib import Path

import torch
import triton
import triton.language as tl


@triton.jit
def ep_scatter_prefix_load_kernel(
    num_recv_tokens_per_expert,
    expert_start_loc,
    m_indices,
    num_experts: tl.constexpr,
    BLOCK_E: tl.constexpr,
    BLOCK_EXPERT_NUM: tl.constexpr,
):
    """Get the current expert offset by reloading the prefix array."""
    cur_expert = tl.program_id(0)

    offset_cumsum = tl.arange(0, BLOCK_EXPERT_NUM)
    tokens_per_expert = tl.load(
        num_recv_tokens_per_expert + offset_cumsum,
        mask=offset_cumsum < num_experts,
        other=0,
    )
    cumsum = tl.cumsum(tokens_per_expert) - tokens_per_expert
    tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)

    cur_expert_start = tl.load(expert_start_loc + cur_expert)
    cur_expert_token_num = tl.load(num_recv_tokens_per_expert + cur_expert)

    m_indices_start_ptr = m_indices + cur_expert_start
    off_expert = tl.arange(0, BLOCK_E)

    for start_m in tl.range(0, cur_expert_token_num, BLOCK_E, num_stages=4):
        tl.store(m_indices_start_ptr + start_m + off_expert, cur_expert)


@triton.jit
def ep_scatter_prefix_sum_kernel(
    num_recv_tokens_per_expert,
    expert_start_loc,
    m_indices,
    num_experts: tl.constexpr,
    BLOCK_E: tl.constexpr,
    BLOCK_EXPERT_NUM: tl.constexpr,
):
    """Get the current expert offset by reducing the values already in registers."""
    cur_expert = tl.program_id(0)

    offset_cumsum = tl.arange(0, BLOCK_EXPERT_NUM)
    tokens_per_expert = tl.load(
        num_recv_tokens_per_expert + offset_cumsum,
        mask=offset_cumsum < num_experts,
        other=0,
    )
    cumsum = tl.cumsum(tokens_per_expert) - tokens_per_expert
    tl.store(expert_start_loc + offset_cumsum, cumsum, mask=offset_cumsum < num_experts)

    cur_expert_start = tl.sum(
        tl.where(offset_cumsum < cur_expert, tokens_per_expert, 0),
        axis=0,
    )
    cur_expert_token_num = tl.load(num_recv_tokens_per_expert + cur_expert)

    m_indices_start_ptr = m_indices + cur_expert_start
    off_expert = tl.arange(0, BLOCK_E)

    for start_m in tl.range(0, cur_expert_token_num, BLOCK_E, num_stages=4):
        tl.store(m_indices_start_ptr + start_m + off_expert, cur_expert)


def compile_and_dump(
    kernel,
    output_path: Path,
    num_experts: int,
    block_e: int,
    num_warps: int,
) -> None:
    block_expert_num = triton.next_power_of_2(num_experts)
    counts = torch.zeros(num_experts, dtype=torch.int32, device="cuda")
    starts = torch.empty(num_experts, dtype=torch.int32, device="cuda")
    m_indices = torch.empty(num_experts * block_e, dtype=torch.int32, device="cuda")

    handle = kernel[(num_experts,)](
        counts,
        starts,
        m_indices,
        num_experts=num_experts,
        BLOCK_E=block_e,
        BLOCK_EXPERT_NUM=block_expert_num,
        num_warps=num_warps,
    )
    output_path.write_text(handle.asm["ptx"])


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--num-experts", type=int, default=24)
    parser.add_argument("--block-e", type=int, default=128)
    parser.add_argument("--num-warps", type=int, default=8)
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path(__file__).resolve().parent / "ep_scatter_ptx",
    )
    args = parser.parse_args()

    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required to compile these Triton kernels")

    args.output_dir.mkdir(parents=True, exist_ok=True)
    compile_and_dump(
        ep_scatter_prefix_load_kernel,
        args.output_dir / "ep_scatter_prefix_load.ptx",
        args.num_experts,
        args.block_e,
        args.num_warps,
    )
    compile_and_dump(
        ep_scatter_prefix_sum_kernel,
        args.output_dir / "ep_scatter_prefix_sum.ptx",
        args.num_experts,
        args.block_e,
        args.num_warps,
    )

if __name__ == "__main__":
    main()

```

Key PTX Code
```
@%p2 st.global.b32 [ %rd4 + 0 ], { %r9 }; // Threads 0-23 store expert_start_loc[lane_id].
	// end inline asm
	.loc	1 33 50                         // compile_ep_scatter_ptx.py:33:50
	mul.wide.u32 	%rd10, %r31, 4;         // cur_expert (ctaid.x) * sizeof(int32).
	add.s64 	%rd5, %rd8, %rd10;          // &expert_start_loc[cur_expert].
	.loc	1 33 31                         // compile_ep_scatter_ptx.py:33:31
	// begin inline asm
	mov.u32 %r10, 0x0;                    // Initialize cur_expert_start.
	ld.global.b32 { %r10 }, [ %rd5 + 0 ]; // All 256 threads load it; no CTA barrier after the store.
```

### Environment

Python:       3.11.12
PyTorch:      2.8.0+cu128
Triton:       3.4.0
GPU:          NVIDIA H20-3e
